### PR TITLE
Rat Dungeon

### PR DESCRIPTION
**Pull request by GarOfMeridian**
This is a cramped warren-style rat dungeon where the rats crawl out of the walls to swarm you. You may have to kill quite a few rats just to fit through and proceed.

Rat king and 8 rat soldiers wait at the bottom, along with a few pieces of interesting low-level loot. When the Rat King dies, all the rats retreat, the floor rises, and you may take the loot and escape.

### CODE REVIEW

File: kod/object/active/holder/room/monsroom/bossroom/ratdungeon.kod [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+
+      propagate;
+   }
+
+%
+%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% End Boss resetting code %
+%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+   ReqLeaveHold(what = $)
```
Comment (skittles1) [inline kod/object/active/holder/room/monsroom/bossroom/ratdungeon.kod]: This message has an error (it should propagate if object not in the boss treasure list) however the exact message you need is already in bossroom.kod so you can just delete this ReqLeaveHold entirely.

File: kod/object/active/holder/room/monsroom/bossroom/ratdungeon.kod [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+
+
+      return;
+   }
+
+   CreateStandardObjects()
+   {
+      local lSkulls;
+
+      lSkulls = [ [13, 13, 15, 36, 30, 26, 32],
```
Comment (skittles1) [inline kod/object/active/holder/room/monsroom/bossroom/ratdungeon.kod]: This list should be constructed with elements of length two [ row, col ] or four [ row, col, finerow, finecol ] so a list with two skulls would look like [ [ row, col, finerow, finecol ] , [ row, col, finerow, finecol ] ].

**Comment by skittles1** [comment]
This is pretty good, some suggestions:
* Make use of the small rats, due to the mob AI they have much more interesting behavior (you'll see what I mean).
* On all rooms you build, make sure you use the "Check" options from the room editor toolbar before saving, in this case to check cross references and remove unused sidedefs.
* This is a newbie dungeon so I think you should add a way to get out if they get stuck in the boss section and can't kill the boss (without logging off), maybe a thin pathway raised a bit off the ground to rationalise the rats not being able to use it (small rats can't reach, soldiers/rat king too big). Given the mob movement upgrades we may have to disallow them specifically in the code, similar to how the lupogg king is prevented from entering too far into the tunnel.

I mentioned a couple errors I found in line notes, these should have been obvious if the room was tested (errors are logged on room creation, and on every monster kill). I know at this point I'm beating a dead horse, but please test before (and I mean before, not after) you make the pull request. By not testing what you're doing is saying "I'm not gonna bother making sure this works, Delerium will do it for me" and I will start taking this personally and closing pulls that are obviously untested. If you were the one testing content for merging, I'm sure you would expect contributors to have tested their own content first.

**Comment by GarOfMeridian** [comment]
"(errors are logged on room creation, and on every monster kill)."

I did not see anything like this. I even made a video in this dungeon. What errors are you seeing?

" I know at this point I'm beating a dead horse, but please test before (and I mean before, not after) you make the pull request. "

Come on, man. I test every single thing I put up and I run through it and play it. Errors do happen, please stop assuming I am not testing.

" By not testing what you're doing is saying "I'm not gonna bother making sure this works, Delerium will do it for me" "

Or you could put up a list of what needs tested and others can help check them off. I am glad to do that. I don't assume you'll do any testing whatsoever, I assume you'll tell me to make a list, and you don't

File: kod/object/active/holder/room/monsroom/bossroom/ratdungeon.kod [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+      
+      pbGenerateMonsters = FALSE;
+      
+      foreach i in plActive
+      {
+         each_obj = Send(self,@HolderExtractObject,#data=i);
+         if IsClass(each_obj,&GiantRat)
+            OR IsClass(each_obj,&GiantRatSoldier)
+         {
+            Send(each_obj,@Delete);
```
Comment (tsShergill) [inline kod/object/active/holder/room/monsroom/bossroom/ratdungeon.kod]: This feels very "clanky". When boss object disappears, it simply deletes the remaining objects? A person could be fighting a creature midway and see it disappear on them. Unless we set off some sort of trigger event such as an earthquake that kills the rats, this doesn't make a lot of sense to me.

Comment (GarOfMeridian) [inline kod/object/active/holder/room/monsroom/bossroom/ratdungeon.kod]: I thought about that, but the rats haven't died, they've simply 'run away' - and they are in areas inaccessible to the boss pit. Not sure if there's a way to do it better. Do we have actual "Run into holes" (and then despawn) AI? That would be awesome

**Comment by skittles1** [comment]
The very thin tunnels in the rat dungeon need to be at least 48 fineunits wide otherwise players and mobs will get stuck on the polygon boundaries. I've fixed most of those issues when I rewrote the room editor/client BSP stuff but there's a subset of the initial problem areas remaining, mainly affecting small spaces like this. Will continue to try and fix it, but for now the tunnel size should be raised (and probably try it out sliding on the walls to see if you get stuck, kind of guessing with 48).

**Comment by GarOfMeridian** [comment]
Ok, tunnels were widened to 48. also rebased. Do you have any suggestions for aggro code? I think the one thing this is missing is actual swarm behavior out of the 'in-wall spawns'.

**Comment by cyberjunk** [comment]
Hm I just want to point out:
48 KOD fine-units is only the (old and new) object-to-object blocking distance. It's not related to the object-to-room blocking. Second one is using the same values on the server as the client does now, so don't blame me, they look like this:

Object-2-Object: 48 KOD fine-units
Object-2-Room: 15.5 KOD fine-units

As you can see objects are currently allowed to get way closer to walls than to other objects....

**Comment by cyberjunk** [comment]
So the conclusion is like this:
1) If you want an object to be able to pass through somewhere tiny, you gotta make the spot at least 16 fine-units, better a bit more than this.
2) However if you want a monster to be able to get through there, if another monster is already sitting in that tiny spot (get around it). Then you kinda need at least: 2x obj-2-room-min-distance + 1x obj-2-obj-min-distance (+a bit more)

**Comment by GarOfMeridian** [comment]
Wait so what does that mean if we do want the tunnels to be only one character or monster wide?

**Comment by cyberjunk** [comment]
Of course these are only the absolute theoretical limits. You gotta add at least +1, usually quite a bit more to make players actually get through without endlessly trying ...

**Comment by skittles1** [comment]
BTW when I deleted the 1.0.16.0-develop branch, all the pull requests for it were automatically closed. These should be opened again pointing at the develop branch (which shouldn't ever change so this should be the last time something like this happens).

**Pull request closed without merging**